// vidstack@1.1.11/dist/prod/chunks/vidstack-bbf44757.js
import { C as Component, a as setStyle, o as onDispose, v as useContext } from './vidstack-4af9f92b.js';
import { aq as FocusVisibleController, ar as tooltipContext, aa as MenuButton } from './vidstack-e5068f63.js';
import { r as requestScopedAnimationFrame } from './vidstack-e557f714.js';

class ControlsGroup extends Component {
  onAttach(el) {
    setStyle(el, "pointer-events", "auto");
  }
}

class TooltipTrigger extends Component {
  constructor() {
    super();
    new FocusVisibleController();
  }
  onConnect(el) {
    onDispose(
      requestScopedAnimationFrame(() => {
        if (!this.connectScope)
          return;
        this.dc();
        const tooltip = useContext(tooltipContext);
        onDispose(() => {
          const button = this.cc();
          button && tooltip.Yb(button);
        });
      })
    );
  }
  dc() {
    const button = this.cc(), tooltip = useContext(tooltipContext);
    button && tooltip.Xb(button);
  }
  cc() {
    return this.el.firstElementChild?.getAttribute("role") === "button" ? this.el.firstElementChild : this.el;
  }
}

class MenuItem extends MenuButton {
}

export { ControlsGroup as C, MenuItem as M, TooltipTrigger as T };
