// vidstack@1.1.11/dist/prod/vidstack.js
import { G as GroupedLog, T as Thumbnail, S as Slider } from './chunks/vidstack-e5068f63.js';
export { A as ARIAKeyShortcuts, t as AudioProviderLoader, ah as AudioRadioGroup, r as AudioTrackList, W as CaptionButton, am as Captions, ai as CaptionsRadioGroup, ag as ChaptersRadioGroup, N as Controls, D as DefaultAudioLayout, I as DefaultLayout, c as DefaultVideoLayout, X as FullscreenButton, F as FullscreenController, al as Gesture, H as HLSProviderLoader, L as List, $ as LiveButton, j as MEDIA_KEY_SHORTCUTS, i as MediaControls, a as MediaPlayer, M as MediaProvider, h as MediaRemoteControl, a9 as Menu, aa as MenuButton, ad as MenuItems, ab as MenuPortal, Y as MuteButton, Z as PIPButton, U as PlayButton, P as PlayerQueryList, an as Poster, ak as QualityRadioGroup, af as Radio, ae as RadioGroup, g as ScreenOrientationController, _ as SeekButton, a8 as SliderChapters, a1 as SliderController, a4 as SliderPreview, a3 as SliderValue, a2 as SliderVideo, aj as SpeedRadioGroup, p as TextRenderers, q as TextTrackList, ap as ThumbnailsLoader, ao as Time, l as TimeRange, a7 as TimeSlider, R as ToggleButton, O as Tooltip, Q as TooltipContent, v as VideoProviderLoader, V as VideoQualityList, a6 as VolumeSlider, e as canFullscreen, J as defaultLayoutContext, d as formatSpokenTime, f as formatTime, o as getTimeRangesEnd, n as getTimeRangesStart, w as isAudioProvider, y as isHLSProvider, z as isHTMLAudioElement, C as isHTMLMediaElement, B as isHTMLVideoElement, x as isVideoProvider, m as mediaContext, k as mediaState, ac as menuPortalContext, E as sliderContext, a0 as sliderState, s as softResetMediaState, a5 as updateSliderPreviewPlacement, b as useDefaultLayoutContext, K as useDefaultLayoutLang } from './chunks/vidstack-e5068f63.js';
import { D as DOMEvent, k as isString, l as listenEvent, u as useState } from './chunks/vidstack-4af9f92b.js';
export { q as appendTriggerEvent, n as findTriggerEvent, m as hasTriggerEvent, i as isKeyboardClick, t as isKeyboardEvent, r as isPointerEvent, w as walkTriggerEventChain } from './chunks/vidstack-4af9f92b.js';
import { T as TextTrackSymbol } from './chunks/vidstack-b3a8b421.js';
export { a as TextTrack, f as findActiveCue, b as isCueActive, i as isTrackCaptionKind, o as observeActiveTextTrack } from './chunks/vidstack-b3a8b421.js';
export { C as ControlsGroup, M as MenuItem, T as TooltipTrigger } from './chunks/vidstack-bbf44757.js';
export { c as canChangeVolume, a as canOrientScreen, b as canPlayHLSNatively, f as canRotateScreen, d as canUsePictureInPicture, e as canUseVideoPresentation } from './chunks/vidstack-bfe4dd65.js';
import './chunks/vidstack-9f573e3f.js';
import './chunks/vidstack-e557f714.js';

class Logger {
  constructor() {
    this.b = null;
  }
  error(...data) {
    return this.dispatch("error", ...data);
  }
  warn(...data) {
    return this.dispatch("warn", ...data);
  }
  info(...data) {
    return this.dispatch("info", ...data);
  }
  debug(...data) {
    return this.dispatch("debug", ...data);
  }
  errorGroup(title) {
    return new GroupedLog(this, "error", title);
  }
  warnGroup(title) {
    return new GroupedLog(this, "warn", title);
  }
  infoGroup(title) {
    return new GroupedLog(this, "info", title);
  }
  debugGroup(title) {
    return new GroupedLog(this, "debug", title);
  }
  setTarget(newTarget) {
    this.b = newTarget;
  }
  dispatch(level, ...data) {
    return this.b?.dispatchEvent(
      new DOMEvent("vds-log", {
        bubbles: true,
        composed: true,
        detail: { level, data }
      })
    ) || false;
  }
}

class LibASSTextRenderer {
  constructor(loader, config) {
    this.priority = 1;
    this.ja = null;
    this.ia = null;
    this.ka = /(ssa|ass)$/;
    this.loader = loader;
    this.config = config;
  }
  canRender(track) {
    return !!track.src && (isString(track.type) && this.ka.test(track.type) || this.ka.test(track.src));
  }
  attach(video) {
    this.loader().then(async (mod) => {
      this.ja = new mod.default({
        ...this.config,
        video,
        subUrl: this.ia?.src || ""
      });
      listenEvent(this.ja, "ready", () => {
        const canvas = this.ja?.ma;
        if (canvas)
          canvas.style.pointerEvents = "none";
      });
      listenEvent(this.ja, "error", (event) => {
        if (this.ia) {
          this.ia[TextTrackSymbol.na] = 3;
          this.ia.dispatchEvent(
            new DOMEvent("error", {
              trigger: event,
              detail: event.error
            })
          );
        }
      });
    });
  }
  changeTrack(track) {
    if (!track || track.readyState === 3) {
      this.la();
    } else if (this.ia !== track) {
      this.ja?.setTrackByUrl(track.src);
      this.ia = track;
    }
  }
  detach() {
    this.la();
  }
  la() {
    this.ja?.freeTrack();
    this.ia = null;
  }
}

class SliderThumbnail extends Thumbnail {
  onAttach(el) {
    this.Fc = useState(Slider.state);
  }
  Gc() {
    const { duration } = this.v.$state;
    return this.Fc.pointerRate() * duration();
  }
}

export { LibASSTextRenderer, Logger, Slider, SliderThumbnail, Thumbnail };
