// vidstack@1.1.11/dist/dev/chunks/vidstack-eb8d860b.js
import { C as Component, a as setStyle, o as onDispose, v as useContext } from './vidstack-454b84a7.js';
import { aq as FocusVisibleController, ar as tooltipContext, aa as MenuButton } from './vidstack-02e38e8c.js';
import { r as requestScopedAnimationFrame } from './vidstack-063df485.js';

class ControlsGroup extends Component {
  onAttach(el) {
    setStyle(el, "pointer-events", "auto");
  }
}

class TooltipTrigger extends Component {
  constructor() {
    super();
    new FocusVisibleController();
  }
  onConnect(el) {
    onDispose(
      requestScopedAnimationFrame(() => {
        if (!this.connectScope)
          return;
        this._attach();
        const tooltip = useContext(tooltipContext);
        onDispose(() => {
          const button = this._getButton();
          button && tooltip._detachTrigger(button);
        });
      })
    );
  }
  _attach() {
    const button = this._getButton(), tooltip = useContext(tooltipContext);
    button && tooltip._attachTrigger(button);
  }
  _getButton() {
    return this.el.firstElementChild?.getAttribute("role") === "button" ? this.el.firstElementChild : this.el;
  }
}

class MenuItem extends MenuButton {
}

export { ControlsGroup as C, MenuItem as M, TooltipTrigger as T };
