// vidstack@1.1.11/dist/prod/providers/vidstack-hls.js
import { a as TextTrack, T as TextTrackSymbol, l as loadScript, p as preconnect } from '../chunks/vidstack-b3a8b421.js';
import { L as ListSymbol, I as IS_CHROME, i as isHLSSupported } from '../chunks/vidstack-bfe4dd65.js';
import { VideoProvider } from './vidstack-video.js';
import { p as peek, D as DOMEvent, l as listenEvent, e as effect, x as camelToKebabCase, y as isUndefined, k as isString, z as isFunction } from '../chunks/vidstack-4af9f92b.js';
import { Q as QualitySymbol, c as coerceToError } from '../chunks/vidstack-9f573e3f.js';
import { R as RAFLoop } from '../chunks/vidstack-768948c3.js';

const toDOMEventType = (type) => camelToKebabCase(type);
class HLSController {
  constructor(_video) {
    this.ja = null;
    this.Ig = null;
    this.A = {};
    this.B = /* @__PURE__ */ new Set();
    this._ = _video;
  }
  get instance() {
    return this.ja;
  }
  setup(ctor, context) {
    this.Wf = context;
    const isLive = peek(context.$state.streamType).includes("live"), isLiveLowLatency = peek(context.$state.streamType).includes("ll-");
    this.ja = new ctor({
      lowLatencyMode: isLiveLowLatency,
      backBufferLength: isLiveLowLatency ? 4 : isLive ? 8 : void 0,
      renderTextTracksNatively: false,
      ...this.A
    });
    const dispatcher = this.Jg.bind(this);
    for (const event of Object.values(ctor.Events))
      this.ja.on(event, dispatcher);
    this.ja.on(ctor.Events.ERROR, this.Pc.bind(this));
    for (const callback of this.B)
      callback(this.ja);
    context.player.dispatch(new DOMEvent("hls-instance", { detail: this.ja }));
    this.ja.attachMedia(this._);
    this.ja.on(ctor.Events.AUDIO_TRACK_SWITCHED, this.Kg.bind(this));
    this.ja.on(ctor.Events.LEVEL_SWITCHED, this.Lg.bind(this));
    this.ja.on(ctor.Events.LEVEL_LOADED, this.Mg.bind(this));
    this.ja.on(ctor.Events.NON_NATIVE_TEXT_TRACKS_FOUND, this.Ng.bind(this));
    this.ja.on(ctor.Events.CUES_PARSED, this.Og.bind(this));
    context.qualities[QualitySymbol.Ra] = this.Pg.bind(this);
    listenEvent(context.qualities, "change", this.Qg.bind(this));
    listenEvent(context.audioTracks, "change", this.Rg.bind(this));
    this.Ig = effect(this.Sg.bind(this));
  }
  Sg() {
    if (!this.Wf.$state.live())
      return;
    const raf = new RAFLoop(this.Tg.bind(this));
    raf.Cb();
    return raf.Db.bind(raf);
  }
  Tg() {
    this.Wf.$state.liveSyncPosition.set(this.ja?.liveSyncPosition ?? Infinity);
  }
  Jg(eventType, detail) {
    this.Wf.player?.dispatch(new DOMEvent(toDOMEventType(eventType), { detail }));
  }
  Ng(eventType, data) {
    const event = new DOMEvent(eventType, { detail: data });
    let currentTrack = -1;
    for (let i = 0; i < data.tracks.length; i++) {
      const nonNativeTrack = data.tracks[i], init = nonNativeTrack.subtitleTrack ?? nonNativeTrack.closedCaptions, track = new TextTrack({
        id: `hls-${nonNativeTrack.kind}${i}`,
        src: init?.url,
        label: nonNativeTrack.label,
        language: init?.lang,
        kind: nonNativeTrack.kind
      });
      track[TextTrackSymbol.na] = 2;
      track[TextTrackSymbol.Ca] = () => {
        if (track.mode === "showing") {
          this.ja.subtitleTrack = i;
          currentTrack = i;
        } else if (currentTrack === i) {
          this.ja.subtitleTrack = -1;
          currentTrack = -1;
        }
      };
      if (nonNativeTrack.default)
        track.setMode("showing", event);
      this.Wf.textTracks.add(track, event);
    }
  }
  Og(eventType, data) {
    const track = this.Wf.textTracks.getById(`hls-${data.track}`);
    if (!track)
      return;
    const event = new DOMEvent(eventType, { detail: data });
    for (const cue of data.cues) {
      cue.positionAlign = "auto";
      track.addCue(cue, event);
    }
  }
  Kg(eventType, data) {
    const track = this.Wf.audioTracks[data.id];
    if (track) {
      this.Wf.audioTracks[ListSymbol.D](
        track,
        true,
        new DOMEvent(eventType, { detail: data })
      );
    }
  }
  Lg(eventType, data) {
    const quality = this.Wf.qualities[data.level];
    if (quality) {
      this.Wf.qualities[ListSymbol.D](
        quality,
        true,
        new DOMEvent(eventType, { detail: data })
      );
    }
  }
  Mg(eventType, data) {
    if (this.Wf.$state.canPlay())
      return;
    const { type, live, totalduration: duration } = data.details;
    const event = new DOMEvent(eventType, { detail: data });
    this.Wf.delegate.u("stream-type-change", {
      detail: live ? type === "EVENT" && Number.isFinite(duration) ? "live:dvr" : "live" : "on-demand",
      trigger: event
    });
    this.Wf.delegate.u("duration-change", { detail: duration, trigger: event });
    const media = this.ja.media;
    if (this.ja.currentLevel === -1) {
      this.Wf.qualities[QualitySymbol.Qa](true, event);
    }
    for (const track of this.ja.audioTracks) {
      this.Wf.audioTracks[ListSymbol.h](
        {
          id: track.id + "",
          label: track.name,
          language: track.lang || "",
          kind: "main"
        },
        event
      );
    }
    for (const level of this.ja.levels) {
      this.Wf.qualities[ListSymbol.h](
        {
          width: level.width,
          height: level.height,
          codec: level.codecSet,
          bitrate: level.bitrate
        },
        event
      );
    }
    media.dispatchEvent(new DOMEvent("canplay", { trigger: event }));
  }
  Pc(eventType, data) {
    if (data.fatal) {
      switch (data.type) {
        case "networkError":
          this.ja?.startLoad();
          break;
        case "mediaError":
          this.ja?.recoverMediaError();
          break;
        default:
          this.ja?.destroy();
          this.ja = null;
          break;
      }
    }
  }
  Pg() {
    if (this.ja)
      this.ja.currentLevel = -1;
  }
  Qg() {
    const { qualities } = this.Wf;
    if (!this.ja || qualities.auto)
      return;
    this.ja[qualities.switch + "Level"] = qualities.selectedIndex;
    if (IS_CHROME)
      this._.currentTime = this._.currentTime;
  }
  Rg() {
    const { audioTracks } = this.Wf;
    if (this.ja && this.ja.audioTrack !== audioTracks.selectedIndex) {
      this.ja.audioTrack = audioTracks.selectedIndex;
    }
  }
  C() {
    if (this.Wf)
      this.Wf.qualities[QualitySymbol.Ra] = void 0;
    this.ja?.destroy();
    this.ja = null;
    this.Ig?.();
    this.Ig = null;
  }
}

class HLSLibLoader {
  constructor(_lib, _ctx, _callback) {
    this.Ye = _lib;
    this.Wf = _ctx;
    this.Wg = _callback;
    this.Ug();
  }
  async Ug() {
    const callbacks = {
      onLoadStart: this.hf.bind(this),
      onLoaded: this.tf.bind(this),
      onLoadError: this.Vg.bind(this)
    };
    let ctor = await loadHLSScript(this.Ye, callbacks);
    if (isUndefined(ctor) && !isString(this.Ye))
      ctor = await importHLS(this.Ye, callbacks);
    if (!ctor)
      return null;
    if (!ctor.isSupported()) {
      const message = "[vidstack]: `hls.js` is not supported in this environment";
      this.Wf.player.dispatch(new DOMEvent("hls-unsupported"));
      this.Wf.delegate.u("error", { detail: { message, code: 4 } });
      return null;
    }
    return ctor;
  }
  hf() {
    this.Wf.player.dispatch(new DOMEvent("hls-lib-load-start"));
  }
  tf(ctor) {
    this.Wf.player.dispatch(
      new DOMEvent("hls-lib-loaded", {
        detail: ctor
      })
    );
    this.Wg(ctor);
  }
  Vg(e) {
    const error = coerceToError(e);
    this.Wf.player.dispatch(
      new DOMEvent("hls-lib-load-error", {
        detail: error
      })
    );
    this.Wf.delegate.u("error", {
      detail: { message: error.message, code: 4 }
    });
  }
}
async function importHLS(loader, callbacks = {}) {
  if (isUndefined(loader))
    return void 0;
  callbacks.onLoadStart?.();
  if (loader.prototype && loader.prototype !== Function) {
    callbacks.onLoaded?.(loader);
    return loader;
  }
  try {
    const ctor = (await loader())?.default;
    if (ctor && !!ctor.isSupported) {
      callbacks.onLoaded?.(ctor);
    } else {
      throw Error(
        false ? "[vidstack] failed importing `hls.js`. Dynamic import returned invalid constructor." : ""
      );
    }
    return ctor;
  } catch (err) {
    callbacks.onLoadError?.(err);
  }
  return void 0;
}
async function loadHLSScript(src, callbacks = {}) {
  if (!isString(src))
    return void 0;
  callbacks.onLoadStart?.();
  try {
    await loadScript(src);
    if (!isFunction(window.Hls)) {
      throw Error(
        false ? "[vidstack] failed loading `hls.js`. Could not find a valid `Hls` constructor on window" : ""
      );
    }
    const ctor = window.Hls;
    callbacks.onLoaded?.(ctor);
    return ctor;
  } catch (err) {
    callbacks.onLoadError?.(err);
  }
  return void 0;
}

const JS_DELIVR_CDN = "https://cdn.jsdelivr.net";
class HLSProvider extends VideoProvider {
  constructor() {
    super(...arguments);
    this.$$PROVIDER_TYPE = "HLS";
    this.z = null;
    this.x = new HLSController(this.video);
    this.y = `${JS_DELIVR_CDN}/npm/hls.js@^1.0.0/dist/hls${".min.js"}`;
  }
  /**
   * The `hls.js` constructor.
   */
  get ctor() {
    return this.z;
  }
  /**
   * The current `hls.js` instance.
   */
  get instance() {
    return this.x.instance;
  }
  static {
    /**
     * Whether `hls.js` is supported in this environment.
     */
    this.supported = isHLSSupported();
  }
  get type() {
    return "hls";
  }
  get canLiveSync() {
    return true;
  }
  /**
   * The `hls.js` configuration object.
   *
   * @see {@link https://github.com/video-dev/hls.js/blob/master/docs/API.md#fine-tuning}
   */
  get config() {
    return this.x.A;
  }
  set config(config) {
    this.x.A = config;
  }
  /**
   * The `hls.js` constructor (supports dynamic imports) or a URL of where it can be found.
   *
   * @defaultValue `https://cdn.jsdelivr.net/npm/hls.js@^1.0.0/dist/hls.min.js`
   */
  get library() {
    return this.y;
  }
  set library(library) {
    this.y = library;
  }
  preconnect() {
    if (!isString(this.y))
      return;
    preconnect(this.y);
  }
  setup(context) {
    super.setup(context);
    new HLSLibLoader(this.y, context, (ctor) => {
      this.z = ctor;
      this.x.setup(ctor, context);
      context.delegate.u("provider-setup", { detail: this });
      const src = peek(context.$state.source);
      if (src)
        this.loadSource(src);
    });
  }
  async loadSource({ src }) {
    if (!isString(src))
      return;
    this.x.instance?.loadSource(src);
  }
  /**
   * The given callback is invoked when a new `hls.js` instance is created and right before it's
   * attached to media.
   */
  onInstance(callback) {
    const instance = this.x.instance;
    if (instance)
      callback(instance);
    this.x.B.add(callback);
    return () => this.x.B.delete(callback);
  }
  destroy() {
    this.x.C();
  }
}

export { HLSProvider };
